File: PaleoRose/Classes/Graphics/XRGraphicPetal.m
//
//  XRGraphicPetal.m
//  XRose
//
//  Created by Tom Moore on Mon Jan 26 2004.
//
// MIT License
//
// Copyright (c) 2004 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#import "XRGraphicPetal.h"
#import "XRGeometryController.h"

@interface XRGraphicPetal()
@property (readwrite, assign) int petalIncrement;
@property (readwrite, assign) float maxRadius;
@property (readwrite, assign) float percent;
@property (readwrite, assign) int count;
@end

@implementation XRGraphicPetal

-(instancetype)initWithController:(id<GraphicGeometrySource>)controller forIncrement:(int)increment forValue:(NSNumber *)aNumber
{
	if (!(self = [super initWithController:controller])) return nil;
	if(self)
	{
		_petalIncrement = increment;
		_percent = [aNumber floatValue];
		_count = [aNumber intValue];
		[self setDrawsFill:YES];
		[self calculateGeometry];
	}
	return self;
}

-(void)calculateGeometry {
	float radius1, radius2;
	float angle1;
	float angle2;
	float angle3;
	float angle4;
    NSPoint pivotPoint = NSMakePoint(0, 0);
	float size = [self.geometryController sectorSize];
	float start = [self.geometryController startingAngle];
    bool isPercent = [self.geometryController isPercent];
	//step 1. find the angles
    angle1 = [self restrictAngleToACircle:(_petalIncrement * size) + start];
    angle2 = [self restrictAngleToACircle:angle1 + size];

    angle3 =  [self restrictAngleToACircle:360.0 - angle1 + 90.0];
    angle4 =  [self restrictAngleToACircle:360.0 - angle2 + 90.0];

	self.drawingPath = [NSBezierPath bezierPath];

    radius1 = isPercent ? [self.geometryController radiusOfPercentValue:0.0] : [self.geometryController radiusOfCount:0];
    radius2 = isPercent ?  [self.geometryController radiusOfPercentValue:_percent] : [self.geometryController radiusOfCount:_count];
    NSPoint startPoint = NSMakePoint(0.0, radius1);
    NSPoint outerPoint = NSMakePoint(0.0, radius2);

    [self.drawingPath moveToPoint:[self.geometryController rotationOfPoint:startPoint byAngle:angle1]];
    [self.drawingPath lineToPoint:[self.geometryController rotationOfPoint:outerPoint byAngle:angle1]];
    [self.drawingPath appendBezierPathWithArcWithCenter:pivotPoint radius:radius2 startAngle:angle3 endAngle:angle4 clockwise:YES];
    [self.drawingPath lineToPoint:[self.geometryController rotationOfPoint:startPoint byAngle:angle2]];
    [self.drawingPath appendBezierPathWithArcWithCenter:pivotPoint radius:radius1 startAngle:angle4 endAngle:angle3 clockwise:NO];
}

-(NSDictionary *)graphicSettings {
	NSMutableDictionary *parentDict = [NSMutableDictionary dictionaryWithDictionary:[super graphicSettings]];

    NSDictionary *classDict = @{
        XRGraphicKeyGraphicType : GraphicTypePetal,
        XRGraphicKeyPetalIncrement : [self stringFromInt: _petalIncrement],
        XRGraphicKeyMaxRadius      : [self stringFromFloat: _maxRadius],
        XRGraphicKeyPercent        : [self stringFromFloat: _percent],
        XRGraphicKeyCount          : [self stringFromInt: _count]
    };
    [parentDict addEntriesFromDictionary:classDict];
	return parentDict;
}
@end

File: PaleoRose/Classes/Graphics/XRGraphicPetal.swift
//
//  XRGraphicPetal.swift
//  PaleoRose
//
//  Created by Cascade on 2025-08-18.
//
//  MIT License
//
//  Copyright (c) 2004 to present Thomas L. Moore.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

import AppKit

/// A petal graphic drawn using the provided geometry controller.
@MainActor
@objc class XRGraphicPetal: XRGraphic {

    /// Keep the NSBezierPath's line width synced when the lineWidth changes.
    override var lineWidth: Float {
        didSet { drawingPath?.lineWidth = CGFloat(lineWidth) }
    }

    /// Ensure newly assigned paths get the current line width.
    override var drawingPath: NSBezierPath? {
        didSet { drawingPath?.lineWidth = CGFloat(lineWidth) }
    }

    // MARK: - Private State

    private var petalIncrement: Int = 0
    private var maxRadius: Float = 0.0
    private var percent: Float = 0.0
    private var count: Int32 = 0

    // MARK: - Public API

    // No default init; use designated initializer below.

    /// Designated initializer.
    /// - Parameters:
    ///   - controller: The geometry controller used to compute drawing geometry.
    ///   - forIncrement: The petal increment index.
    ///   - forValue: The value for this petal (interpreted as count or percent by the controller).
    @objc
    init?(controller: GraphicGeometrySource, forIncrement increment: Int32, forValue aNumber: NSNumber) {
        super.init(controller: controller)
        petalIncrement = Int(increment)
        percent = aNumber.floatValue
        count = aNumber.int32Value
        drawsFill = true
        calculateGeometry()
    }

    private func calculateAngles(petalIncrement: Int32, size: CGFloat, start: CGFloat) -> [CGFloat] {
        [
            CGFloat(
                restrictAngle(
                    toACircle: Float(CGFloat(petalIncrement) * size + start)
                )
            ),
        ]
    }

    /// Recalculate geometry when external changes require it.
    @objc override func calculateGeometry() {
        guard let controller = geometryController else { return }

        // step 1. find the angles
        let size = CGFloat(controller.sectorSize())
        let start = CGFloat(controller.startingAngle())

        let angle1Polar = (CGFloat(petalIncrement) * size + start).normalizePositiveAngle()
        let angle2Polar = (angle1Polar + size).normalizePositiveAngle()

        let angle1Canvas = angle1Polar.cgAngleFromPolar().normalizePositiveAngle()
        let angle2Canvas = angle2Polar.cgAngleFromPolar().normalizePositiveAngle()

        let isPercent = controller.isPercent()

        let radius1: CGFloat = isPercent ? CGFloat(controller.radius(ofPercentValue: 0.0)) : CGFloat(
            controller.radius(ofCount: 0)
        )
        let radius2: CGFloat = isPercent ? CGFloat(controller.radius(ofPercentValue: Double(percent))) : CGFloat(
            controller.radius(ofCount: count)
        )

        let pivotPoint = CGPoint(x: 0, y: 0)
        let startPoint = CGPoint(x: 0.0, y: radius1)
        let outerPoint = CGPoint(x: 0.0, y: radius2)

        let path = NSBezierPath()
        path.lineWidth = CGFloat(lineWidth)
        path
            .move(
                to: controller.rotation(of: startPoint, byAngle: Double(angle1Polar))
            )
        path
            .line(
                to: controller.rotation(of: outerPoint, byAngle: Double(angle1Polar))
            )
        path.appendArc(withCenter: pivotPoint, radius: radius2, startAngle: angle1Canvas, endAngle: angle2Canvas, clockwise: true)
        path
            .line(
                to: controller.rotation(of: startPoint, byAngle: Double(angle2Polar))
            )
        path.appendArc(withCenter: pivotPoint, radius: radius1, startAngle: angle2Canvas, endAngle: angle1Canvas, clockwise: false)

        drawingPath = path
    }

    /// Returns the graphic's settings merged with superclass settings.
    @objc override func graphicSettings() -> [AnyHashable: Any] {
        var parent = super.graphicSettings()
        // Keys match those used throughout the ObjC code and tests.
        let classDict: [AnyHashable: Any] = [
            "GraphicType": "Petal",
            "_petalIncrement": string(from: Int32(petalIncrement)),
            "_maxRadius": string(from: maxRadius),
            "_percent": string(from: percent),
            "_count": string(from: Int32(count))
        ]
        parent.merge(classDict) { _, new in new }
        return parent
    }
}
